In [3]:
import pandas as pd

# Final demo dataset

In [5]:
df = pd.read_csv("../Data/Raw/df_final_demo.csv")
df.head()

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0     836976             6.0              73.0      60.5     U        2.0   
1    2304905             7.0              94.0      58.0     U        2.0   
2    1439522             5.0              64.0      32.0     U        2.0   
3    1562045            16.0             198.0      49.0     M        2.0   
4    5126305            12.0             145.0      33.0     F        2.0   

         bal  calls_6_mnth  logons_6_mnth  
0   45105.30           6.0            9.0  
1  110860.30           6.0            9.0  
2   52467.79           6.0            9.0  
3   67454.65           3.0            6.0  
4  103671.75           0.0            3.0

In [ ]:
len(df)

70609

In [ ]:
df.tail()

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
70604    7993686             4.0              56.0      38.5     U        3.0   
70605    8981690            12.0             148.0      31.0     M        2.0   
70606     333913            16.0             198.0      61.5     F        2.0   
70607    1573142            21.0             255.0      68.0     M        3.0   
70608    5602139            21.0             254.0      59.5     F        3.0   

              bal  calls_6_mnth  logons_6_mnth  
70604  1411062.68           5.0            5.0  
70605   101867.07           6.0            6.0  
70606    40745.00           3.0            3.0  
70607   475114.69           4.0            4.0  
70608   157498.73           7.0            7.0

In [ ]:
df.describe()

client_id  clnt_tenure_yr  clnt_tenure_mnth      clnt_age  \
count  7.060900e+04    70595.000000      70595.000000  70594.000000   
mean   5.004992e+06       12.052950        150.659367     46.442240   
std    2.877278e+06        6.871819         82.089854     15.591273   
min    1.690000e+02        2.000000         33.000000     13.500000   
25%    2.519329e+06        6.000000         82.000000     32.500000   
50%    5.016978e+06       11.000000        136.000000     47.000000   
75%    7.483085e+06       16.000000        192.000000     59.000000   
max    9.999839e+06       62.000000        749.000000     96.000000   

          num_accts           bal  calls_6_mnth  logons_6_mnth  
count  70595.000000  7.059500e+04  70595.000000   70595.000000  
mean       2.255528  1.474452e+05      3.382478       5.566740  
std        0.534997  3.015087e+05      2.236580       2.353286  
min        1.000000  1.378942e+04      0.000000       1.000000  
25%        2.000000  3.734683e+04      1.000000       4.000000  
50%        2.000000  6.333290e+04      3.000000       5.000000  
75%        2.000000  1.375449e+05      6.000000       7.000000  
max        8.000000  1.632004e+07      7.000000       9.000000

In [ ]:
df["client_id"].duplicated().sum()

np.int64(0)

In [ ]:
df['gendr'].value_counts() 

gendr
U    24122
M    23724
F    22746
X        3
Name: count, dtype: int64

In [ ]:
# SABINA COMMENT: Why not do df.isna().sum()?
df['gendr'].isnull().sum()

np.int64(14)

In [ ]:
df['clnt_tenure_yr'].isnull().sum()

np.int64(14)

In [ ]:
df['clnt_tenure_mnth'].isnull().sum()

np.int64(14)

In [ ]:
df['clnt_age'].isnull().sum()

np.int64(15)

In [ ]:
df['num_accts'].isnull().sum()

np.int64(14)

In [ ]:
df['bal'].isnull().sum()

np.int64(14)

In [ ]:
df['calls_6_mnth'].isnull().sum()

np.int64(14)

In [ ]:
df['logons_6_mnth'].isnull().sum()

np.int64(14)

In [9]:
df[df.isnull().any(axis=1)]


client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
4164     7402828             NaN               NaN       NaN   NaN        NaN   
8316      355337             NaN               NaN       NaN   NaN        NaN   
8677     8412164             NaN               NaN       NaN   NaN        NaN   
9583     4666211             8.0             106.0       NaN     F        2.0   
13444    2222915             NaN               NaN       NaN   NaN        NaN   
18066    4876926             NaN               NaN       NaN   NaN        NaN   
25961    5277910             NaN               NaN       NaN   NaN        NaN   
28432    7616759             NaN               NaN       NaN   NaN        NaN   
35323    8191345             NaN               NaN       NaN   NaN        NaN   
43518    1227228             NaN               NaN       NaN   NaN        NaN   
46076    8611797             NaN               NaN       NaN   NaN        NaN   
47189    5144725             NaN               NaN       NaN   NaN        NaN   
49846    1037867             NaN               NaN       NaN   NaN        NaN   
63701    1618749             NaN               NaN       NaN   NaN        NaN   
66219    6021001             NaN               NaN       NaN   NaN        NaN   

            bal  calls_6_mnth  logons_6_mnth  
4164        NaN           NaN            NaN  
8316        NaN           NaN            NaN  
8677        NaN           NaN            NaN  
9583   42550.55           4.0            7.0  
13444       NaN           NaN            NaN  
18066       NaN           NaN            NaN  
25961       NaN           NaN            NaN  
28432       NaN           NaN            NaN  
35323       NaN           NaN            NaN  
43518       NaN           NaN            NaN  
46076       NaN           NaN            NaN  
47189       NaN           NaN            NaN  
49846       NaN           NaN            NaN  
63701       NaN           NaN            NaN  
66219       NaN           NaN            NaN

In [6]:
df['num_accts'].min(), df['num_accts'].max()

(1.0, 8.0)

In [10]:
df['clnt_age'].min(), df['clnt_age'].max()

(13.5, 96.0)

In [11]:
df['calls_6_mnth'].min(), df['calls_6_mnth'].max()

(0.0, 7.0)

# Experiment clients dataset

In [7]:
df1 = pd.read_csv("../Data/Raw/df_final_experiment_clients.csv")
df1.head()

client_id Variation
0    9988021      Test
1    8320017      Test
2    4033851   Control
3    1982004      Test
4    9294070   Control

In [ ]:
len(df1)

70609

In [ ]:
df["client_id"].duplicated().sum()

np.int64(0)

In [ ]:
df1["Variation"].value_counts()

Variation
Test       26968
Control    23532
Name: count, dtype: int64

In [ ]:
df1['Variation'].isna().sum()


np.int64(20109)

# Webdata dataset - Part-1 

SABINA COMMENT: Would've advised you to analyze both of these (pt1 and pt2) together, I don't see any reason why they should be separate.

In [8]:
df2 = pd.read_csv("../Data/Raw/df_final_web_data_pt_1.csv")
df2.head()

client_id            visitor_id                      visit_id process_step  \
0    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   
1    9988021  580560515_7732621733  781255054_21935453173_531117       step_2   
2    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   
3    9988021  580560515_7732621733  781255054_21935453173_531117       step_2   
4    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   

             date_time  
0  2017-04-17 15:27:07  
1  2017-04-17 15:26:51  
2  2017-04-17 15:19:22  
3  2017-04-17 15:19:13  
4  2017-04-17 15:18:04

In [ ]:
len(df2)

343141

In [ ]:
df2["process_step"].unique()

array(['step_3', 'step_2', 'step_1', 'start', 'confirm'], dtype=object)

In [ ]:
df2["process_step"].value_counts()

process_step
start      108910
step_1      73432
step_2      61768
step_3      53628
confirm     45403
Name: count, dtype: int64

In [ ]:
df2["client_id"].value_counts()

client_id
2128341    84
9800867    66
9638063    64
2313292    61
5165430    60
           ..
3939374     1
8477848     1
2606984     1
5141668     1
1266421     1
Name: count, Length: 58391, dtype: int64

In [ ]:
df2['client_id'].nunique()

58391

In [ ]:
steps_per_client = (
    df2.groupby('client_id')['process_step']
    .nunique()
    .sort_values(ascending=False)
)
steps_per_client.describe()

count    58391.000000
mean         3.974585
std          1.526914
min          1.000000
25%          3.000000
50%          5.000000
75%          5.000000
max          5.000000
Name: process_step, dtype: float64

- Median is 5, so at least 50% of the clients completed the process
- Most of the clients completed the process (75% is 5)   
SABINA COMMENT: Nope, 75% in the `describe()` method is a point (Q3), not an interval (min-Q3). The conclusion you had before was right and the reason that one is correct is because having 5 at the Q2 point means everything >Q2 (50% of the data) will be higher than 5.
75% (Q3) being equal to 5 means that everything >Q3 is >5 => 25% of the data is surely higher than 5. DM me if still unclear on this.

- Mean is 3.97, denotes that some clients dropped off during the process 

# Webdata dataset - Part-2

In [ ]:
df3 = pd.read_csv("Data/Raw/df_final_web_data_pt_2.csv")
df3.head()

client_id             visitor_id                      visit_id  \
0     763412  601952081_10457207388  397475557_40440946728_419634   
1    6019349  442094451_91531546617  154620534_35331068705_522317   
2    6019349  442094451_91531546617  154620534_35331068705_522317   
3    6019349  442094451_91531546617  154620534_35331068705_522317   
4    6019349  442094451_91531546617  154620534_35331068705_522317   

  process_step            date_time  
0      confirm  2017-06-06 08:56:00  
1      confirm  2017-06-01 11:59:27  
2       step_3  2017-06-01 11:58:48  
3       step_2  2017-06-01 11:58:08  
4       step_1  2017-06-01 11:57:58

In [ ]:
len(df3)

412264

In [ ]:
df3["process_step"].unique()

array(['confirm', 'step_3', 'step_2', 'step_1', 'start'], dtype=object)

In [ ]:
df3["process_step"].value_counts()

process_step
start      135035
step_1      89761
step_2      71294
step_3      58614
confirm     57560
Name: count, dtype: int64

In [ ]:
df3["client_id"].value_counts()

client_id
5574879    104
2274658     90
5951379     80
1952790     80
3545814     72
          ... 
4220991      1
7506474      1
7414592      1
1156842      1
674799       1
Name: count, Length: 67430, dtype: int64

In [ ]:
df3['client_id'].nunique()

67430

In [ ]:
steps_per_client = (
    df3.groupby('client_id')['process_step']
    .nunique()
    .sort_values(ascending=False)
)
steps_per_client.describe()

count    67430.000000
mean         4.025463
std          1.494016
min          1.000000
25%          3.000000
50%          5.000000
75%          5.000000
max          5.000000
Name: process_step, dtype: float64

- Median is 5, so at least 50% of the clients completed the process
- Most of the clients completed the process (75% is 5)
- Mean is 4.02, denotes that some clients dropped off during the later stages of the process